src/navigation/failure_identification/watchdog.py

```diff
@@ -1,11 +1,97 @@
 import numpy as np
 import rospy
 from pandas import DataFrame
+from util.ros_utils import get_rosparam
+from util.SO3 import SO3
+from typing import Tuple
+
+WINDOW_SIZE = get_rosparam("watchdog/window_size", 100)
+ANGULAR_THRESHOLD = get_rosparam("watchdog/angular_threshold", 0.001)
+LINEAR_THRESHOLD = get_rosparam("watchdog/linear_threshold", 0.55)
 
 
 class WatchDog:
-    def __init__(self):
-        pass
+    def __init__(self, collector_in):
+        self.collector = collector_in
+
+    def get_start_end_positions(self, dataframe: DataFrame) -> Tuple[np.ndarray, np.ndarray]:
+        # get the average of the first 25% of the dataframes x, y, z position from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
+        end = dataframe.loc[dataframe["row"] >= len(dataframe) - cutoff]
+        start_x = np.mean(start["x"])
+        start_y = np.mean(start["y"])
+        start_z = np.mean(start["z"])
+        start_pos = np.array([start_x, start_y, start_z])
+        end_x = np.mean(end["x"])
+        end_y = np.mean(end["y"])
+        end_z = np.mean(end["z"])
+        end_pos = np.array([end_x, end_y, end_z])
+        return start_pos, end_pos
+
+    def get_start_end_rotations(self, dataframe: DataFrame) -> Tuple[SO3, SO3]:
+        # get the average of the first 25% of the dataframes rotation from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
+        end = dataframe.loc[dataframe["row"] >= len(dataframe) - cutoff]
+        start_rot = np.array(
+            [np.mean(start["rot_x"]), np.mean(start["rot_y"]), np.mean(start["rot_z"]), np.mean(start["rot_w"])]
+        )
+        end_rot = np.array([np.mean(end["rot_x"]), np.mean(end["rot_y"]), np.mean(end["rot_z"]), np.mean(end["rot_w"])])
+        return SO3(start_rot), SO3(end_rot)
+
+    def get_start_end_time(self, dataframe: DataFrame) -> Tuple[float, float]:
+        # get the average of the first 25% of the dataframes time from the inputted dataframe
+        cutoff = int(len(dataframe) * 0.25)
+        start = dataframe.loc[dataframe["row"] <= cutoff]
+        end = dataframe.loc[dataframe["row"] >= len(dataframe) - cutoff]
+        start_time = np.mean(start["time"])
+        end_time = np.mean(end["time"])
+        return start_time, end_time
+
+    def check_angular_stuck(self, delta_time, delta_rot, dataframe) -> bool:
+        # check to make sure all cmd_vel_twist values are the same sign and cmd_vel_x values are 0
+        turn_are_all_same_sign = dataframe["cmd_vel_twist"].apply(np.sign).eq(dataframe["cmd_vel_twist"].iloc[0]).all()
```

Comment:
what if angular is zero? doesn't this still return true?

Reply:
yea thats why we check that they're not all zero

Reply:
where? doesn't the line below this only check if the linear commanded velocity is zero, not the angular?

Reply:
ah mb yea this is still fine because the is_stuck function is only called when the state is not done or off meaning we are moving and either the linear velocity or the turning velocity must be non zero

Reply:
sounds good